### PR TITLE
Use forEach when it makes sense.

### PR DESCRIPTION
**Pull request by n1k0**
I'm finding all these for loops hard to read and error prone since we use `var` instead of `let` statements.

I'm suggesting we loop using `forEach` whenever it makes sense to address both issues.

@leibovic @gijsk thoughts?

### CODE REVIEW

File: Readability.js [hunk 1]
```diff
@@ -119,6 +119,36 @@ Readability.prototype = {
   },
 
   /**
+   * Iterate over a NodeList, which doesn't natively fully implement the Array
+   * interface.
```
Comment (DavidBruant) [inline Readability.js]: [NodeLists](https://dom.spec.whatwg.org/#interface-nodelist) are `iterable<node>`.
This means they can be iterated over with for..of loops.
It could result in the following code:
````js
function convertRelativeURIs(tagName, propName) {
    for(var elem of articleContent.getElementsByTagName(tagName)){
        var relativeURI = elem.getAttribute(propName);
        if (relativeURI != null)
            elem.setAttribute(propName, toAbsoluteURI(relativeURI));
    }
}
````js

This would work in Firefox, but wouldn't in Node 0.10. It can be made to work in Node 0.12 and io.js with some experimental flags on I believe. The future (hopefully coming soon) is that ES6 will be deployed consistently everywhere, so it will eventually work everywhere.

I don't know what's best, but felt it's worth sharing options :-)

Comment (n1k0) [inline Readability.js]: Readability's code needs to run on iOS, so I suspect what you suggest is not an option (sadly).

File: Readability.js [hunk 2]
```diff
@@ -1058,7 +1067,6 @@ Readability.prototype = {
   _findNextPageLink: function(elem) {
     var uri = this._uri;
     var possiblePages = {};
-    var allLinks = elem.getElementsByTagName('a');
```
Comment (gijsk) [inline Readability.js]: Err, this is used a few lines down... I mean, the code isn't currently being called at all, but we should try not to break it like this. :-)

Comment (n1k0) [inline Readability.js]: Fixed, I've been a little too prone at cutting things it seems :)

I've added more tests to cover things I've refactored.

File: Readability.js (hunk 1 above)
Comment (gijsk) [inline Readability.js]: The downside of forEach used to be that the extra function calls produced overhead, so the performance is worse. I don't know if that's still (meaningfully) the case.

File: Readability.js [hunk 3]
```diff
@@ -217,19 +246,17 @@ Readability.prototype = {
     var doc = this._doc;
 
     // Remove all style tags in head
-    var styleTags = doc.getElementsByTagName("style");
-    for (var st = styleTags.length - 1; st >= 0; st -= 1) {
-      styleTags[st].parentNode.removeChild(styleTags[st]);
-    }
+    this._forEachNode(doc.getElementsByTagName("style"), function(styleNode) {
+      styleNode.parentNode.removeChild(styleNode);
+    });
```
Comment (gijsk) [inline Readability.js]: Certainly here though, the issue I'd be worried about is that Array.forEach can't deal with removing items from the array while iterating. The reason this loop iterated backwards is to deal with NodeLists that are dynamic, ie when you call removeChild, the item will disappear from the doc.getElementsByTagName result, and therefore from the nodelist.

Your change does not handle that, so on compliant DOM implementations it would break this code. It doesn't with JSDOMParser, but I believe at least iOS will want to use Readability with a "real" DOM.

File: Readability.js (hunk 1 above)
Comment (n1k0) [inline Readability.js]: I'm always prepending test suite run calls with the `time` exec, I didn't notice any performance impact. I think js engines are now optimizing this properly, and even if it doesn't perform as good I think more readable code is worth the price.

Comment (DavidBruant) [inline Readability.js]: Functions calls are removed by the JIT compiler, especially if the collection type is stable.
If you notice a significant difference (which is unlikely), file a bug, that's the sort of things JS engine folks want to know about.
More on the topic http://rfrn.org/~shu/2013/03/20/two-reasons-functional-style-is-slow-in-spidermonkey.html

Comment (gijsk) [inline Readability.js]: > Functions calls are removed by the JIT compiler, especially if the collection type is stable.

Which isn't the case here for real DOMs, right? :-)

File: Readability.js (hunk 3 above)
Comment (n1k0) [inline Readability.js]:
```js
var a = [1, 2, 3];
a.forEach(function(x, i) {
  if (x === 2)
    a.splice(i, 1)
});
console.log(a) // [1,3]
```

I agree this isn't the most easy code to deal with, but in my world still better than within traditional for loops (I've probably been too much into FP latesly, trying to avoid mutable state changes as much as I can, or when I can't trying to limit side effects as much as possible :/)

Comment (n1k0) [inline Readability.js]: Hmm, you're probably right that dealing with a mutable DOM that way is probably not a good idea. I should just cancel this PR then.

File: Readability.js (hunk 1 above)
Comment (n1k0) [inline Readability.js]: Guys, please note we're speaking milliseconds here. I think the argument of the impossibility to deal with DOM mutations within function based loops is much more worth spending our time.

File: Readability.js (hunk 3 above)
Comment (gijsk) [inline Readability.js]: Hm, no, actually the docs say that I am just wrong:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach

> The range of elements processed by forEach() is set before the first invocation of callback. Elements that are appended to the array after the call to forEach() begins will not be visited by callback. If the values of existing elements of the array are changed, the value passed to callback will be the value at the time forEach() visits them; elements that are deleted before being visited are not visited.

I don't think any of these loops care about added items?

Comment (n1k0) [inline Readability.js]: > I don't think any of these loops care about added items?

It doesn't seem so, though that means we won't ever be able to introduce such a behavior; on the other hand, appending items to a currently iterated collection is definitely asking for trouble imoe :D

File: test/test-pages/style-tags-removal/source.html [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+      Tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam,
+      quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo
+      consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse
+      cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non
+      proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
+    </div>
+  </article>
+  <style>
+    * {
+      color: yelow;
```
Comment (gijsk) [inline test/test-pages/style-tags-removal/source.html]: Nit: yellow with two ls

File: test/test-pages/basic-tags-cleaning/source.html [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+      cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non
+      proident, sunt in culpa qui officia deserunt mollit anim id est laborum.</p>
+    </div>
+    <h2>Foo</h2>
+    <div>
+      <p>Tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam,
+      quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo
+      consequat.</p>
+      <object data="foo.swf" type="application/x-shockwave-flash" width="88" height="31">
+        <param movie="foo.swf" />
```
Comment (gijsk) [inline test/test-pages/basic-tags-cleaning/source.html]: Can you add a youtube example to this that is kept, please?

File: Readability.js [hunk 6]
```diff
@@ -1571,7 +1579,7 @@ Readability.prototype = {
     if (!metadata.excerpt) {
       var paragraphs = articleContent.getElementsByTagName("p");
       if (paragraphs.length > 0) {
```
Comment (gijsk) [inline Readability.js]: Huh, can you file a followup issue to make this less dumb? If we only care about the first P, we should just do our own iteration.

File: Readability.js [hunk 7]
```diff
@@ -978,16 +987,16 @@ Readability.prototype = {
...
-    for (var i = 0, il = links.length; i < il; i += 1) {
-      linkLength += this._getInnerText(links[i]).length;
-    }
+    // XXX implement _reduceNodeList?
+    this._forEachNode(element.getElementsByTagName("a"), function(linkNode) {
+      linkLength += this._getInnerText(linkNode).length;
+    });
 
-    return linkLength / textLength;
+    return textLength !== 0 ? linkLength / textLength : 0;
```
Comment (gijsk) [inline Readability.js]: Why was this necessary?

File: Readability.js [hunk 8]
```diff
@@ -915,13 +924,13 @@ Readability.prototype = {
...
    * @param Element
+   * @param Boolean normalizeSpaces (default: true)
    * @return string
   **/
   _getInnerText: function(e, normalizeSpaces) {
     var textContent = e.textContent.trim();
-    normalizeSpaces = (typeof normalizeSpaces === 'undefined') ? true : normalizeSpaces;
 
-    if (normalizeSpaces) {
+    if (!Boolean(normalizeSpaces)) {
```
Comment (gijsk) [inline Readability.js]: Ewww. No need to convert.

File: Readability.js [hunk 9]
```diff
@@ -900,14 +912,11 @@ Readability.prototype = {
...
-    for (var i = 0; i < length; i++) {
-      var child = e.children[i];
-      if (this.DIV_TO_P_ELEMS.indexOf(child.tagName) !== -1 || this._hasChildBlockElement(child))
-        return true;
-    }
-    return false;
+  _hasChildBlockElement: function (element) {
+    return this._someNode(element.childNodes, function(node) {
+      return this.DIV_TO_P_ELEMS.indexOf(node.tagName) !== -1 ||
+             this._hasChildBlockElement(node)
```
Comment (gijsk) [inline Readability.js]: Nit: semicolon.

File: Readability.js [hunk 10]
```diff
@@ -860,14 +878,13 @@ Readability.prototype = {
    * @param Element
   **/
   _removeScripts: function(doc) {
-    var scripts = doc.getElementsByTagName('script');
-    for (var i = scripts.length - 1; i >= 0; i -= 1) {
-      scripts[i].nodeValue="";
-      scripts[i].removeAttribute('src');
+    this._forEachNode(doc.getElementsByTagName('script'), function(scriptNode) {
+      scriptNode.nodeValue = "";
```
Comment (gijsk) [inline Readability.js]: I wonder why we set `nodeValue` at all, and why we nullcheck the element's `parentNode`... Well, best leave that as-is for now, and file another followup, please? :-)

File: Readability.js [hunk 11]
```diff
@@ -544,17 +563,17 @@ Readability.prototype = {
        * A score is determined by things like number of commas, class names, etc. Maybe eventually link density.
       **/
       var candidates = [];
-      for (var pt = 0; pt < elementsToScore.length; pt += 1) {
-        var parentNode = elementsToScore[pt].parentNode;
+      this._forEachNode(elementsToScore, function(elementToScore) {
```
Comment (gijsk) [inline Readability.js]: So, here's something which I've been wondering about... are you trying to be consistent for its own sake? Because you could just call `elementsToScore.forEach(...)`, right? Or update `_forEachNode` to check `Array.isArray` before doing the `.call` thing...

File: Readability.js [hunk 12]
```diff
@@ -336,26 +361,21 @@ Readability.prototype = {
...
-        articleParagraphs[i].parentNode.removeChild(articleParagraphs[i]);
-    }
+    this._forEachNode(articleContent.getElementsByTagName('p'), function(paragraph) {
+      var imgCount = paragraph.getElementsByTagName('img').length;
+      var embedCount = paragraph.getElementsByTagName('embed').length;
+      var objectCount = paragraph.getElementsByTagName('object').length;
+      var totalCount = imgCount + embedCount + objectCount;
+
+      if (totalCount === 0 && !this._getInnerText(paragraph, false))
+        paragraph.parentNode.removeChild(paragraph);
```
Comment (gijsk) [inline Readability.js]: Hrmpf. All of this is pretty inefficient, but that can be another followup...

File: Readability.js (hunk 7 above)
Comment (n1k0) [inline Readability.js]: Possible division by zero. Defensive programming, admittedly.

File: Readability.js (hunk 3 above)
Comment (gijsk) [inline Readability.js]: Hrmpf. I thought this was all going to go OK, and then I realized we'll still need to fix the `_setNodeTag` stuff for non-JSDOMParser, which will involve reparenting stuff (ie, below, move everything from a `font` tag into a different `span` tag and then replace the `font` with the `span`). This is unfortunate because if there are nested tags, we will be adding + removing things, which will mess up the loop. For instance:

`<font><font>Hello</font></font>`

We'll hit the first font first with the forEach loop, which then means we'll reparent the second one, which *might* mean we won't iterate over the second node? Not sure... :-\

Tested in Firefox's devtools console:
```
>>> document.getElementsByTagName("font").forEach
undefined
>>> Array.prototype.forEach.call(document.getElementsByTagName("font"), function(f) { var x = document.createElement("span"); Array.prototype.forEach.call(f.childNodes, function(fc) { x.appendChild(fc); }); f.parentNode.replaceChild(x, f); });
undefined
>>> document.getElementsByTagName("font")
HTMLCollection [ <font> ]
>>> document.getElementsByTagName("font")[0].parentNode
<span>
```

File: Readability.js (hunk 7 above)
Comment (gijsk) [inline Readability.js]: I see. In that case, can we just return early before calling getElementsByTagName if `textLength === 0`? If that is the case, clearly `linkLength` is also going to be 0 anyway...

File: Readability.js (hunk 11 above)
Comment (n1k0) [inline Readability.js]: > Or update _forEachNode to check Array.isArray before doing the .call thing...

Would this be to address a possible performance penalty? Because applying the method from the prototype on an array works exactly the same way, and has nearly no impact on performances…

Comment (gijsk) [inline Readability.js]: No, I think I was just confused by the fact that we're using this helper function all the time, which says it's there to make things work with NodeList, but we use it for other things, too... :-)

I don't really mind either way, up to you!

File: Readability.js (hunk 12 above)
Comment (n1k0) [inline Readability.js]: Heh. You can't believe the wtf-per-minute I've been producing while initially reading the code. We'll address all this iteratively, if you don't mind.

Comment (gijsk) [inline Readability.js]: Yeah, it was the same on my side. And yes, iteratively for sure. See also https://twitter.com/mleibovic/status/578676569582907392 for the state of some of the people here...

File: Readability.js (hunk 7 above)
Comment (n1k0) [inline Readability.js]: I like the idea of avoiding calling expensive functions for no reason :) Will fix.

**Comment by gijsk** [comment]
@n1k0 I need to head out and likely won't be back until the evening here in SF.

I think this is basically OK but we need to be careful with the forEach stuff in cases where we manipulate the things we're looping. Can you just file a followup that as well? I think the clarity this change adds is something we should take while we're still working on the compat with live DOMs.

**Comment by gijsk** [comment]
(by which I mean, we can postpone the hard work of figuring out those changes later, and potentially change them back to manual loops if that fixes bugs at that point in time)

File: Readability.js (hunk 3 above)
Comment (n1k0) [inline Readability.js]: Damn. Sounds so much complicated. Indirectly related, just out of curiosity: why do we want to replace font tags with spans? couldn't we just remove these font tags? Note that we currently keep the font tag attributes, which makes generated HTML looking a little silly.

Comment (gijsk) [inline Readability.js]: It's to avoid the styles of the readable page being messed up by the page's (font tag's) styles, I think. Like I said, we can leave it for a followup, it's not currently broken with this change.

File: test/test-pages/basic-tags-cleaning/source.html (hunk 5 above)
Comment (n1k0) [inline test/test-pages/basic-tags-cleaning/source.html]: Good point, especially as youtube embedded code is now iframe-based only.

Comment (n1k0) [inline test/test-pages/basic-tags-cleaning/source.html]: Filed https://github.com/mozilla/readability/issues/56

File: Readability.js (hunk 10 above)
Comment (n1k0) [inline Readability.js]: Filed https://github.com/mozilla/readability/issues/57

File: Readability.js (hunk 6 above)
Comment (n1k0) [inline Readability.js]: I don't get why this code is a problem. Hints?

**Comment by n1k0** [comment]
> Can you just file a followup that as well? 
https://github.com/mozilla/readability/issues/59

File: Readability.js (hunk 6 above)
Comment (gijsk) [inline Readability.js]: It gets all the paragraphs (walks the entire tree), but then we only really want the first one. It could just stop looking after the first paragraph, ie the equivalent of querySelector. :-)

**Comment by gijsk** [comment]
Great, this is ready to be merged now, right, @n1k0 ?

**Comment by n1k0** [comment]
> Great, this is ready to be merged now, right, @n1k0 ?

I think so! :)

File: Readability.js (hunk 6 above)
Comment (gijsk) [inline Readability.js]: Filed #60

**Comment by n1k0** [comment]
\o/
